### ISSUE TITLE
How To

### ISSUE THREAD

**Issue by ModynAI** [elyha7/yoloface#1]
Hi,
I can not use your Repo for inference.

I have 'cloned' your repo, 'cd' to it, then installed dependencies which has been mentioned in 'requirements.txt'
and then used 
````
from yoloface import YoloDetector
import numpy as np
from PIL import Image

model = YoloFace(target_size=720,gpu=0,min_face=90)
orgimg = np.array(Image.open('test_image.jpg'))
bboxes,points = model.predict(orgimg)

````

But it returns that yoloface is not installed!

>  ---------------------------------------------------------------------------
> ImportError                               Traceback (most recent call last)
> <ipython-input-11-82275d1c6223> in <module>()
> ----> 1 from yoloface import YoloDetector
>       2 import numpy as np
>       3 from PIL import Image
>       4 
>       5 model = YoloFace(target_size=720,gpu=0,min_face=90)
> 
> /content/yoloface/yoloface.py in <module>()
>      11 from math import sqrt
>      12 
> ---> 13 from .models.common import Conv
>      14 from .models.yolo import Model
>      15 from .utils.datasets import letterbox
> 
> ImportError: attempted relative import with no known parent package
> 
> ---------------------------------------------------------------------------
> NOTE: If your import is failing due to a missing package, you can
> manually install dependencies using either !pip or !apt.
> 
> --------------------------------------------------------------------------

**How should i fix this?**

Thanks
Best regards

**Comment by elyha7** [comment]
Changed repo's structure to use absolute imports, now your problem must be solved. If you want to import module outside the repo's folder, export path into PYTHONPATH (example in readme).

**Comment by ModynAI** [comment]
its Now have diffrent problem!

```
from face_detector import YoloDetector
import numpy as np
from PIL import Image

model = YoloDetector(target_size=720,gpu=-1,min_face=90)
orgimg = np.array(Image.open('/content/13.jpg'))
bboxes,points = model.predict(orgimg)
```

> -1
> /content/yoloface/weights/yolov5n_state_dict.pt
> ---------------------------------------------------------------------------
> AttributeError                            Traceback (most recent call last)
> <ipython-input-12-84e37c16c307> in <module>()
>       4 
>       5 model = YoloDetector(target_size=720,gpu=-1,min_face=90)
> ----> 6 orgimg = np.array(Image.open('/content/13.jpg'))
>       7 bboxes,points = model.predict(orgimg)
> 
> 4 frames
> /usr/local/lib/python3.7/dist-packages/PIL/TiffImagePlugin.py in ImageFileDirectory_v2()
>     667                 (TiffTags.FLOAT, "f", "float"),
>     668                 (TiffTags.DOUBLE, "d", "double"),
> --> 669                 (TiffTags.IFD, "L", "long"),
>     670             ],
>     671         )
> 
> AttributeError: module 'PIL.TiffTags' has no attribute 'IFD'



I've used -1 for using CPU, but even in GPU=0 the same happen

**Comment by ModynAI** [comment]
Thanks
Its fixed Now. but i dont have output image

its just says that 

> 0
> /content/yoloface/weights/yolov5n_state_dict.pt
> WARNING: --img-size 612 must be multiple of max stride 32, updating to 640


How can i see the Bounding box position in the picture?

Thanks 
Best regards

**Comment by ModynAI** [comment]
Thanks to you, i can have the bboxes and points but is it possible to use it on webcam? as a real time face detection?

Thanks 
Best regards

**Comment by elyha7** [comment]
> Thanks to you, i can have the bboxes and points but is it possible to use it on webcam? as a real time face detection?
> 
> Thanks Best regards

Yeah, you can. Follow this guide, but use my detector instead of Haar cascade.
https://realpython.com/face-detection-in-python-using-a-webcam/

**Comment by ModynAI** [comment]
Thank you,

Have you tries to add TENSORT for the model as there is in 
yolov5-face

**Comment by AndreKev** [comment]
> 

How did you proceed to draw the boxes ?

Think it can be done by simply drawing boxes in bbox. Alright ?

**Comment by elyha7** [comment]
> How did you proceed to draw the boxes ?
> 
> Think it can be done by simply drawing boxes in bbox. Alright ?

You can use PIL, opencv or any other drawing library to draw bboxes and facial points. Here is an example using opencv.
```python
for box,lm in zip(bboxes,points):
    x1,y1,x2,y2 = box
    orgimg = cv2.rectangle(orgimg,(x1,y1),(x2,y2),(255,0,0),3)
    for i in lm:
        x = i[0]
        y = i[1]
        orgimg = cv2.circle(orgimg, (x, y), 3, (0,255,0), -1)
```

**Comment by AndreKev** [comment]
i got an issue unpacking box. 

This is how I fixed this 

```
for box,lm in zip(bboxes,points):
    for x1,y1,x2,y2 in box :
        count += 1
        orgimg = cv2.rectangle(orgimg,(x1,y1),(x2,y2),(255,0,0),3)
        for i in lm:
            x = i[0][0]
            y = i[0][1]
            #orgimg = cv2.circle(orgimg, (x, y), 3, (0,255,0), 1)
```

I get an issue drawing the circle. x and y are not numbers but lists.

What do they represent ?

**Comment by AndreKev** [comment]
Furthermore, it seems like this model detects better with PIL images than with cv2 images.
Loading the linked image with cv2 gives 5 recognized faces, and loading with PIL gives 7 faces.

Here is the image: ![48](https://user-images.githubusercontent.com/86581549/168311616-058b931f-f275-44d0-9bf3-f701e82bf3a0.jpeg)

loading the image with PIL (your loading technique) 7 faces: 
![new](https://user-images.githubusercontent.com/86581549/168312624-fa7b4797-189d-4e8c-8715-811b30aee2eb.jpeg)


Loading the image with cv2 5 faces:
![new](https://user-images.githubusercontent.com/86581549/168312252-0691776b-3abd-4989-b6a8-89b67d1f356b.jpeg)


I think it is all about colormode. You can consider let me sending a pull request to update drawing engine.

Congratulations. Do you know how I can improve my detection ?

**Comment by AndreKev** [comment]
I just changes the minimum face size with PIL image rendering:

Result :
![new](https://user-images.githubusercontent.com/86581549/168316117-6709f9b3-5422-45c7-aa8d-f2426e5f198c.jpeg)


Awesome ! But still to be improved since some faces aren't shaded

**Comment by AndreKev** [comment]
But I notice that in RGB some faces are detected and not in BGR, then vice versa !

**Comment by elyha7** [comment]
> But I notice that in RGB some faces are detected and not in BGR, then vice versa !

The models were trained with rgb images and opencv by default uses bgr color space, so ofc better accuracy in rgb. In opencv color space can be changed with cv2.cvtColor() method. If you need better accuracy for smaller faces, consider using other models from [deepcam-cn](https://github.com/deepcam-cn/yolov5-face) repo, i use one of the lightest ones in my repo.

**Comment by AndreKev** [comment]
> 

Alright but look at this image I link. You will notice that, some face are only detected in BGR (blue boxes) and not in RGB, then some face are only detected in RGB (red boxes) and not in BGR.

From hidden experience I think there is greater probability on detecting dark faces and faces with mufflers in BGR, but face detection in BGR outputs a little more false positives in these cases.


![84_detected](https://user-images.githubusercontent.com/86581549/168560256-40591e37-d40f-4800-92c0-d3fa0088ec61.JPG)
